profiling: add internal changes to support profiling of gRPC

adtac opened the pull request.

dfawley commented:
We'll have to make `vet.sh` happy even though we only quasi-support appengine standard now.  Can you add the exclude-appengine build tag to anything importing `unsafe`?

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,213 @@
+// InitStats initialises all the relevant Stat objects. Must be called exactly
+// once per lifetime of a process; calls after the first one are ignored.
+func InitStats(streamStatsSize uint32) (err error) {
+	if !atomic.CompareAndSwapInt32(&statsInitialised, 0, 1) {
```
How about using a sync.One to perform initialization that you want to be done only one?

adtac replied:
I think a CAS is much cleaner and simpler than defining a func and calling sync.Once. I'll admit I didn't know about sync.Once before, but that looks more useful when there are multiple concurrent goroutines contending for the same thing.

Looking at [the source](https://golang.org/src/sync/once.go#L40), they even talk about not using a CAS because they have different goals -- we explicitly want the function to do nothing and return when things are already initialised.

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,213 @@
+// pointer) and the user is expected to lose their reference to the timer to
+// allow the Timer object to be garbage collected.
+func (stat *Stat) AppendTimer(timer *Timer) (index int) {
+	if stat != nil && timer != nil {
```
Why are you concerned about `stat == nil` in all these methods? When can that happen?

adtac replied:
If stats are disabled, `http2_server.go` and `http2_client.go` will not create a new `Stat` object at stream initialisation. However, we have calls to stuff like `stream.Stat().NewTimer()` everywhere even though `stream.Stat()` returns nil. This is so that we can have one-line API to hook stats everywhere. That is, adding just one line to the top of a function will automatically do everything:

```go
defer stream.Stat().NewTimer("whatever").Egress()
```

NewTimer will return nil because `stream.Stat()` is nil, and as a result `Egress`, when it is executed at function executed, will also do nothing.

This is simpler than what stat handler and channelz do with ugly `if channelz.IsOn() { ... }` multi-line statements everywhere.

adtac replied:
See the hooks PR for examples: https://github.com/grpc/grpc-go/pull/3159

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,213 @@
+		stat.mu.Unlock()
+	}
+
+	return
```
Usually prefer handling error conditions upfront. So, this would become:

```golang
if stat == nil {
  return
}
// Do things which need to be done when stat != nil

```

adtac replied:
Done.

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,213 @@
+// new timer every time. Instead a slice of timers large enough to hold most
+// data is created and elements from this slice are used. This should put less
+// pressure on the GC too.
+func (stat *Stat) NewTimer(timerTag string) (index int) {
```
Named return value parameters don't lead to very readable code. Do you have a strong reason to use them here?

Excerpt from internal Go readability doc:
```
Don't name result parameters just to avoid declaring a var inside the function; that trades off a minor implementation brevity at the cost of unnecessary API verbosity.

Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

Finally, in some cases you need to name a result parameter in order to change it in a deferred closure. That is always OK.
```

adtac replied:
Not really, I don't have a real reason. I agree that it's much more readable to declare the variables inside the function.

Done.

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,213 @@
+
+// A Timer represents the wall-clock beginning and ending of a logical
+// operation.
+type Timer struct {
```
The `Timer` struct has the same fields (functionally) as that of the `TimerProto`. Would it make sense to not define this type at all, and just use the protos here as well?

adtac replied:
I did that originally, but I faced two disadvantages with using the same struct:

 - `TimerProto` has a few extra fields, which amount to at least 8-10 additional bytes. We want to keep the memory footprint of profiling be as low as possible.

 - `time.Time` does not translate well to protobuf-lingo. There's a time struct that can be imported from Google's protobuf extensions, but internally, it's basically a `sec` field and a `nsec` field, which we end up doing any way. To translate from `time.Now()` (local time) to a UTC sec, nsec pair, we need to do more computation at `Timer` creation time (and `Egress`, of course), which we'd like to avoid.

adtac replied:
Currently, it takes about ~50 bytes to store a stat. Using `TimerProto` would effectively mean 16% fewer stats for the same amount of memory used.

adtac replied:
Resolved offline.

adtac commented:
Pushed V2 addressing review comments.

Ping @dfawley @easwars

dfawley commented:
Travis is unhappy right now:
```
internal/profiling/profiling.go:149:13: undeclared name: goId (compile)
internal/profiling/profiling.go:149:3: unknown field GoId in struct literal (compile)
internal/profiling/profiling.go:149:13: undeclared name: goId (compile)
internal/profiling/profiling.go:149:3: unknown field GoId in struct literal (compile)
internal/profiling/profiling.go:149:13: undeclared name: goId (compile)
internal/profiling/profiling.go:149:3: unknown field GoId in struct literal (compile)
```

adtac commented:
ugh, fixed (waiting for travis)

adtac commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,257 @@
+	"unsafe"
+)
+
+type queue struct {
```
Done.

adtac commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,257 @@
+		arr:      make([]unsafe.Pointer, size),
+		size:     size,
+		mask:     size - 1,
+		acquired: 0,
```
Done.

adtac commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,257 @@
+
+// Allocates and returns a new *queue. size needs to be a exponent of two.
+func newQueue(size uint32) (q *queue) {
+	q = &queue{
```
Done. I really don't want to code golf haha, this is a remnant from a few weeks ago when I weirdly became obsessed with named return variables (bad idea).

adtac commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,257 @@
+	qp.q0 = unsafe.Pointer(newQueue(size))
+	qp.q1 = unsafe.Pointer(newQueue(size))
+	qp.q = qp.q0
+	return
```
Done.

adtac commented:
@dfawley thanks for the review, addressed your comments and pushed V7, V8. (But don't bother with the actual commits; I imagine you're already using the Files Changed tab.)

easwars commented on internal/profiling/buffer_test.go:
```diff
@@ -0,0 +1,188 @@
+	cb, err := newCircularBuffer(size)
+	if err != nil {
+		t.Errorf("error allocating CircularBuffer: %v", err)
+		return
```
Here and in other places where it does not make sense for the test to continue, use `t.Fatal` instead of `t.Error` followed by `return`

adtac replied:
Done.

easwars commented on internal/profiling/buffer_test.go:
```diff
@@ -0,0 +1,188 @@
+
+	result = cb.Drain()
+	if uint32(len(result)) != size/2 {
+		t.Errorf("expected result size %d, got %d", size/2, len(result))
```
Error message should usually have the following format:
```foobar(args) = <whatever you got>, want <whatever you want>```

For example, this is an excerpt from the godoc for the testing package:
```golang
func TestAbs(t *testing.T) {
    got := Abs(-1)
    if got != 1 {
        t.Errorf("Abs(-1) = %d; want 1", got)
    }
}
```

adtac replied:
Done.

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,220 @@
+	StatTag  string
+	Timers   []Timer
+	Metadata []byte
+	mu       sync.Mutex
```
Either move the `mu` to the top of the struct, thereby implying that all the fields following it are protected by it, or add a comment saying which fields are protected by this mutex.

adtac replied:
Done.

easwars commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,220 @@
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation.
+const defaultStatAllocatedTimers int32 = 128
```
What is the rationale behind the value of 128?

adtac replied:
A typical unary request creates about 80-100 timers. This is mostly anecdotal and may vary depending on different factors, but this is the average number I've seen in the past few weeks. Should I include this in the comment?

easwars replied:
Sure, a comment should be helpful for the future.

easwars commented on internal/profiling/profiling_test.go:
```diff
@@ -0,0 +1,45 @@
+	"sync"
+	"testing"
+)
+
```
Do you plan on having any tests for the exported functions from `profiling.go`?

adtac replied:
Done.

adtac commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,271 @@
+	for i := from; i < to; i++ {
+		// We have mutual exclusion on arr, there's no need for atomics.
+		x := (*interface{})(arr[i])
+		if x != nil {
```
When `Push` is executed, if the `postDrainingCheck` check fails, `acquired` is decremented and the whole process is retried by getting a potentially different `acquired` value. This means that `acquired % size` wasn't actually written to.

As a result, in some very rare cases, the array pointers may be nil.

adtac replied:
I think I can explain this much better in person; let me know if you have time for a quick five minute chat by the whiteboard.

easwars replied:
Makes sense. A small comment for the same would be useful. Thanks.

adtac commented:
@easwars thanks for the comments, pushed `PATCH V9`, PTAL.

dfawley commented:
FYI travis is failing; it looks like your test needs to be conditional on appengine?

adtac commented:
Ah yes, I forgot to add the build tag to the new tests. Pushed V10.

dfawley commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,252 @@
+
+	stat.mu.Lock()
+	if index < len(stat.Timers) {
+		stat.Timers[index].End = time.Now()
```
Should we call time.Now() before taking the lock?

adtac replied:
Yes, good point. While there is a case to be made that this might call time.Now wastefully even if the index check fails, that should ideally never happen. Done.

dfawley commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,252 @@
+		GoID:     goid(),
+	})
+	index := len(stat.Timers) - 1
+	stat.mu.Unlock()
```
Do we need to do all of this inside the lock?  Would it be more efficient to do:
```go
t := Timer{}
stat.mu.Lock()
stat.Timers=append(stat.Timers, t)
index := len(stat.Timers)-1
stat.mu.Unlock()
// set t.TimerTag, t.Begin, and t.GoID
```

adtac replied:
It's also more accurate to do it this way -- time spent on lock acquisition and release isn't included in the difference.

dfawley commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,252 @@
+	// reallocation and copy, which is probably unnecessary code complexity if
+	// the improvement isn't significant.
+	stat.mu.Lock()
+	stat.Timers = append(stat.Timers, Timer{
```
Would it be more efficient to do something like this (does this avoid stack allocations and copies or does the Go compiler already effectively do this?):

```go
var t *Timer
l := len(stat.Timers)
if l < cap(stat.Timers) {
  stat.Timers=stat.Timers[:l+1]
} else {
  stat.Timers=append(stat.Timers, Timer{})
}
t=&stat.Timers[l]
// Fill in t's fields
```

adtac replied:
I benchmarked something like this before and there was no discernible performance improvement/difference, so I stuck to the more maintainable and readable version (this one).

adtac commented on internal/profiling/goid_modified.go:
```diff
@@ -0,0 +1,68 @@
+//    and it takes about ~2 ns/op, which is negligible in the face of the tens
+//    of microseconds that grpc takes to complete a RPC request.
+//
+// A personal note (adtac@): I'm fully aware of the Go team's position on why
```
Done.

adtac commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,252 @@
+// The profiling package contains two logical components: buffer.go and
+// profiling.go. The former implements a circular buffer (a.k.a. ring buffer)
+// in a lock-free manner using atomics. This ring buffer is used by
+// profiling.go to store various statistics. For example, StreamStats is a
```
there is a `internal/buffer` that was recently created by @easwars - perhaps that? It could house all kinds of generic data buffer-related data structures.

adtac replied:
Currently it just contains one thing -- an unbounded channel-like data structure.

adtac replied:
Moving to a new internal/profiling/buffer package, resolved offline.

adtac commented:
@dfawley thanks for the review, addressed all but two comments (unresolved). PTAL.

menghanl commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,270 @@
+	// After the completion of a Push operation, the written counter is
+	// incremented. Also used by drainWait to wait for all pushes to complete.
+	written uint32
+	// The Drain operation must first choose a queue to push to. Once this is
```
The __Push__ operation?

menghanl commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,270 @@
+		// Reverse our increment and retry. Since there's no SubUint32 in atomic,
+		// ^uint32(0) is used to denote -1.
+		atomic.AddUint32(&q.acquired, ^uint32(0))
+		goto reloadq
```
Nit: rewrite to `for { if break }`?

menghanl commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,270 @@
+
+	// At this point, we're definitely writing to the right queue. That is, one
+	// of the following is true:
+	//   1. No drainer is in execution.
```
A drainer is draining the other queue? (Essentially same as no drainer, though..)

menghanl commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,270 @@
+	wg.Add(int(numCircularBufferPairs))
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		go func(qi uint32) {
+			qs[qi].drainWait()
```
Can this happen?

| Push                                | Drain                                      |
|-------------------------------------|--------------------------------------------|
| load q                              |                                            |
|                                     | drainWait(), returns                       |
| q.acquired++                        |                                            |
| q.drainingPostCheck == 0, not retry |                                            |
|                                     | q.drainingPostCheck = 1 (this is too late) |

Push will write to this queue, while Drain is draining it, right?

adtac replied:
Excellent catch! I've changed the algorithm slightly to do the barrier check before the switching happens.

menghanl commented on internal/profiling/buffer.go:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		qs[i].acquired = 0
```
These need to be atomic, too, I believe.
There's a chance that Push() is accessing those concurrently (if the Push() goroutine is not scheduled after loading q, until this)

adtac replied:
Done.

dfawley commented on internal/profiling/goid_modified.go:
```diff
@@ -52,17 +52,30 @@ import (
 //    (a quadratic number of variants to maintain).
 //
-// 4. This approach, which requires a simple two-line modification to the Go
+// 4. This approach, which requires a simple two-line modification [1] to the Go
```
Hey, that's four lines!  ;)

Seriously though, it's super minor but please delete "two-line" -- "simple modification" and a reference is sufficient.

adtac replied:
Done.

adtac commented:
@dfawley, @menghanl thanks for the review! Updated the PR:

1. Rebased on to latest master.
2. reverted the lock-free stat/timer thing -- I found a data race in that approach with some additional testing. Apparently, dynamically resizing, lock-free arrays are apparently an area of active research for a reason :P 
3. Addressed review comments.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+// Note that it is impossible to enable profiling for one server and leave it
+// turned off for another. This is intentional and by design -- if the status
+// of profiling was server-specific, clients wouldn't be able to profile
+// themselves. As a result, SetEnabled turns profiling on and off for all
```
s/SetEnabled/Enable

adtac replied:
Done.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+// servers and clients in the binary. Each stat will be, however, tagged with
+// whether it's a client stat or a server stat; so you should be able to filter
+// for the right type of stats in post-processing.
+// SetEnabled is the internal
```
s/SetEnabled/Enable

Or just delete this line

adtac replied:
Note sure where that line came from haha. Deleted.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+	// End marks the end of a timer.
+	End time.Time
+	// Each Timer must be started and ended within the same goroutine; GoID
+	// captures this goroutine ID. The Go runtime doesn not typically expose this
```
s/doesn/does

adtac replied:
Done.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+type Timer struct {
+	// Each Timer has a tag that identifies it uniquely within a Stat. This is
+	// usually a forward-slash-separated hierarchical string.
+	TimerTag string
```
Call this just `Tag`?

If renaming causes too much trouble, we can keep this.
There's also a rename tool: https://godoc.org/golang.org/x/tools/cmd/gorename

adtac replied:
Never underestimate the power of `:%s/TimerTag/Tags/g` :P 

Done.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+type Stat struct {
+	// Each Stat has a tag associated with it to identify it in post-processing.
+	// This is can be any arbitrary string.
+	StatTag string
```
Same as timer, just `Tag`.
We can keep it if renaming isn't trivial.

adtac replied:
Done.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
```
`NewTimer` isn't allocation free, and does allocate a `Timer`, right?

Did you mean to `make([]Timer, defaultStatAllocatedTimers)` in `NewStat`, and set the fields of a timer in `NewTimer`? (This makes `defaultStatAllocatedTimers` a hard limit of the slice length, though)

Or this was the old behavior, you changed the implementation, but forgot to update the comment?

adtac replied:
Yeah, I previous used to do `Timers[i].field = value` with a fixed size Timer array.

But I did some benchmarks and I was surprised to find that doing `append(slice, &Timer{})` is faster than `append(slice, Timer{})` -- I wrongly assumed that `append(slice, Timer{})` would be faster. Changed the API slightly to use a `*Timer`.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+// ServerConnectionCounter counts the number of connections a server has seen.
+// This counter is embedded within a StreamStat's Metadata along with each
+// stream's stream ID to uniquely identify a query. Accessed atomically.
+var ServerConnectionCounter uint64
```
I'm assuming this is `++` by the server, and then read by the server (or the transport). And the stat metadata is `<connectionID><streamID>`.

We should make this an internal detail of the transport I think.
Connection counter should be the same as transport counter (because transport and connection are 1:1 mapping).
For the metadata field, we can treat it as an opaque stream ID generated by transport, and is guaranteed to be unique. The transport can still generate it the same way `<transportID><streamID>`.

adtac replied:
Agreed, done.

menghanl commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,251 @@
+// initiated. This counter is embedded within a StreamStat's Metadata along
+// with each stream's stream ID to uniquely identify a query. Accessed
+// atomically.
+var ClientConnectionCounter uint64
```
Similar as server connection counter.

adtac replied:
Agreed, done.

adtac commented on internal/profiling/buffer/buffer.go:
```diff
@@ -0,0 +1,268 @@
+	// that any existing drainWaits can proceed.
+	if atomic.LoadUint32(&cb.qp[n].drainExecuting) != 0 {
+		atomic.AddUint32(&q.written, 1)
+		return
```
I did change it to a `for { if break }`, but changed it again to this.

No, the caller wouldn't call Push again. This because dropping some pushes is okay if it guarantees finite completion of the push operation. This would fail only if a Drain starts right at the start of a Push, which is pretty rare, but if it happens, we'd still like for profiling to not affect gRPC performance.

adtac commented on internal/profiling/buffer/buffer.go:
```diff
@@ -0,0 +1,268 @@
+	// We say "B has wrapped around A" when this happens. In this case, since A
+	// occurred before B, B's Push should be the final value. However, we
+	// accommodate A being the final value because wrap-arounds are extremely
+	// rare and accounting for them requires an additional counter and a
```
>The only guarantee is that one of A or B is kept, but it's not always B, right?

Yes, not always.

>And depending on the queue size, this is not always "extremely rare". If we want to keep this rare, we should keep the size above some minimum value (it's currently 8, I'm not sure if that's big enough)

CircularBuffer averages ~77.3ns/op per Push operation (one slot per RPC). Let's say a Push operation has started, but is being wrapped around by a different Push operation. Then, gRPC must have handled at least 8 queries within 77ns, which suggests a QPS of 100M+ req/s, which is currently impossible, no?

adtac commented on internal/profiling/buffer/buffer.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+// Push pushes an element in to the circular buffer.
+func (cb *CircularBuffer) Push(x interface{}) {
```
There is no guarantee. Push chooses to drop stats and overwrite them in the interest of keeping impact of profiling low. A similar data structure with locks that is deterministic and guarantees that the newest `size` objects will be kept costs ~1us per RPC, which is non-negligible.

adtac commented on internal/profiling/buffer/buffer.go:
```diff
@@ -0,0 +1,268 @@
+
+	qs := make([]*queue, len(cb.qp))
+	for i := 0; i < len(cb.qp); i++ {
+		atomic.StoreUint32(&cb.qp[i].drainExecuting, 1)
```
I ended up removing drainExecuting altogether -- a simple queue pointer check is actually better.

adtac commented on internal/profiling/profiling.go:
```diff
@@ -0,0 +1,222 @@
+type Timer struct {
+	// Each Timer has a tag that identifies it uniquely within a Stat. This is
+	// usually a forward-slash-separated hierarchical string.
+	Tags string
```
I just changed it from `s/TimerTag/Tags`, but since it's a comma-separated list of strings, so I feel Tags is justified here.

adtac replied:
But I should update the documentation a bit, yes.

dfawley replied:
SGTM

adtac commented on internal/profiling/buffer/buffer_appengine.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func (cb *CircularBuffer) Drain() []interface{} {
+	return make([]interface{}, 0)
```
Done.

adtac commented:
@menghanl pushed V19 modifying CircularBuffer's documentation.
